crates/turborepo-auth/src/auth/login.rs

```diff
@@ -0,0 +1,334 @@
+const DEFAULT_PORT: u16 = 9789;
+
+/// Login writes a token to disk at token_path. If a token is already present,
+/// we do not overwrite it and instead log that we found an existing token.
+pub async fn login<F>(
+    api_client: &impl Client,
+    ui: &UI,
+    existing_token: Option<&str>,
+    mut set_token: F,
+    login_url_configuration: &str,
+    login_server: &impl LoginServer,
+) -> Result<()>
+where
+    F: FnMut(&str) -> Result<()>,
+{
+    // Check if token exists first.
+    if let Some(token) = existing_token {
+        if let Ok(response) = api_client.get_user(&token).await {
+            println!("{}", ui.apply(BOLD.apply_to("Existing token found!")));
+            ui::print_cli_authorized(&response.user.email, ui);
+            return Ok(());
+        }
+    }
+
+    let redirect_url = format!("http://{DEFAULT_HOST_NAME}:{DEFAULT_PORT}");
+    let mut login_url = Url::parse(login_url_configuration)?;
+
+    login_url
+        .path_segments_mut()
+        .map_err(|_: ()| Error::LoginUrlCannotBeABase {
+            value: login_url_configuration.to_string(),
+        })?
+        .extend(["turborepo", "token"]);
+
+    login_url
+        .query_pairs_mut()
+        .append_pair("redirect_uri", &redirect_url);
+
+    println!(">>> Opening browser to {login_url}");
+    let spinner = start_spinner("Waiting for your authorization...");
+
+    // Try to open browser for auth confirmation.
+    {
+        let url = login_url.as_str();
+        if webbrowser::open(url).is_err() {
+            warn!("Failed to open browser. Please visit {url} in your browser.");
+        }
+    };
+
+    let token_cell = Arc::new(OnceCell::new());
+    login_server
+        .run(
+            DEFAULT_PORT,
+            login_url_configuration.to_string(),
+            token_cell.clone(),
+        )
+        .await?;
+
+    spinner.finish_and_clear();
+
+    let token = token_cell
+        .get()
+        .ok_or_else(|| anyhow!("Failed to get token"))?;
+
+    // This function is passed in from turborepo-lib
+    // TODO: inline this here and only pass in the location to write the token as an
+    // optional arg.
+    set_token(token)?;
+
+    // TODO: make this a request to /teams endpoint instead?
+    let user_response = api_client.get_user(token.as_str()).await?;
+
+    ui::print_cli_authorized(&user_response.user.email, ui);
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::atomic::AtomicUsize;
+
+    use async_trait::async_trait;
+    use reqwest::{Method, RequestBuilder, Response};
+    use turborepo_api_client::{Client, Error, Result};
+    use turborepo_vercel_api::{
+        CachingStatusResponse, Membership, PreflightResponse, Role, SpacesResponse, Team,
+        TeamsResponse, User, UserResponse, VerifiedSsoUser,
+    };
+    use turborepo_vercel_api_mock::start_test_server;
+
+    use super::*;
+
+    // Used to track how many times the server was hit.
+    lazy_static::lazy_static! {
```

Comment:
Could we store this in the `MockLoginServer` struct?

Reply:
Yep, not a problem. I'll move it in there - that was the initial purpose anyway 🔥